packages/hedera/src/dids/HederaDidRegistrar.ts

```diff
@@ -0,0 +1,182 @@
+import {
+  AgentContext,
+  DidCreateResult,
+  DidDeactivateResult,
+  DidDocument,
+  DidDocumentKey,
+  DidDocumentRole,
+  DidDocumentService,
+  DidRecord,
+  DidRegistrar,
+  DidRepository,
+  DidUpdateResult,
+  JsonTransformer,
+} from '@credo-ts/core'
+import {
+  HederaDidCreateOptions,
+  HederaDidDeactivateOptions,
+  HederaDidUpdateOptions,
+  HederaLedgerService,
+} from '../ledger/HederaLedgerService'
+
+export class HederaDidRegistrar implements DidRegistrar {
+  public readonly supportedMethods = ['hedera']
+
+  async create(agentContext: AgentContext, options: HederaDidCreateOptions): Promise<DidCreateResult> {
+    try {
+      const didRepository = agentContext.dependencyManager.resolve(DidRepository)
+      const ledgerService = agentContext.dependencyManager.resolve(HederaLedgerService)
+
+      const { did, didDocument, rootKey } = await ledgerService.createDid(agentContext, options)
+
+      const credoDidDocument = new DidDocument({
+        ...didDocument,
+        service: didDocument.service?.map((s) => new DidDocumentService(s)),
+      })
+
+      await didRepository.save(
+        agentContext,
+        new DidRecord({
+          did,
+          role: DidDocumentRole.Created,
+          didDocument: credoDidDocument,
+          keys: [rootKey],
+        })
+      )
+
+      return {
+        didDocumentMetadata: {},
+        didRegistrationMetadata: {},
+        didState: {
+          state: 'finished',
+          did,
+          didDocument: credoDidDocument,
+        },
+      }
+    } catch (error) {
+      agentContext.config.logger.debug('Error creating DID', {
+        error,
+      })
+      return {
+        didDocumentMetadata: {},
+        didRegistrationMetadata: {},
+        didState: {
+          state: 'failed',
+          reason: `Unable to register Did: ${error.message}`,
+        },
+      }
+    }
+  }
+
+  async update(agentContext: AgentContext, options: HederaDidUpdateOptions): Promise<DidUpdateResult> {
+    const didRepository = agentContext.dependencyManager.resolve(DidRepository)
+    const ledgerService = agentContext.dependencyManager.resolve(HederaLedgerService)
+
+    try {
+      const { did } = options
+      const { didDocument, didDocumentMetadata } = await ledgerService.resolveDid(agentContext, did)
+      const didRecord = await didRepository.findCreatedDid(agentContext, did)
+      if (!didDocument || didDocumentMetadata.deactivated || !didRecord) {
+        return {
+          didDocumentMetadata: {},
+          didRegistrationMetadata: {},
+          didState: {
+            state: 'failed',
+            reason: 'Did not found',
+          },
+        }
+      }
+
+      const keys = this.concatKeys(didRecord.keys, options.secret?.keys)
```

Comment:
Do we handle removal of keys? Do we keep them linked in the did record still?

Reply:
Good point as well, I added removal of non-relevant keys after DID update operation.